### PR TITLE
[improvement] Provide span operation when using deferred traces

### PR DESCRIPTION
**Pull request by pkoenig10**
## Before this PR
When using versions of tracing-java with https://github.com/palantir/tracing-java/pull/97, the spans created by the concurrency limiter and the dispatcher have the same name as the parent span. This is confusing since the parent span name does not accurately describe the work of the concurrency limiter or the dispatcher.

## After this PR
The concurrency limiter and the dispatcher provide a unique span operation when creating spans.

### CODE REVIEW

File: okhttp-clients/src/main/java/com/palantir/conjure/java/okhttp/IoExceptionResponseHandler.java [hunk 1]
```diff
@@ -44,7 +44,7 @@
                     : "<empty>";
 
             return Optional.of(new SafeIoException(
-                    String.format("Error %s. (Failed to parse response body as SerializableError.)", response.code()),
+                    "Failed to parse response body as SerializableError",
```
Comment (pkoenig10) [inline okhttp-clients/src/main/java/com/palantir/conjure/java/okhttp/IoExceptionResponseHandler.java:47]: This needed to change as a result of upgrading `safe-logging` which requires these messages to be compile-time constants.

File: okhttp-clients/src/main/java/com/palantir/conjure/java/okhttp/OkHttpClients.java [hunk 2]
```diff
@@ -73,7 +73,7 @@
      * have not been used for sixty seconds are terminated, so this won't block shutdown indefinitely.
      */
     private static final ExecutorService executionExecutor =
-            Tracers.wrap(Executors.newCachedThreadPool(executionThreads));
+            Tracers.wrap("dispatcher", Executors.newCachedThreadPool(executionThreads));
```
Comment (carterkozak) [inline okhttp-clients/src/main/java/com/palantir/conjure/java/okhttp/OkHttpClients.java]: nit: Can we provide more context by prefixing these with `OkHttp: `?

File: okhttp-clients/src/main/java/com/palantir/conjure/java/okhttp/RemotingOkHttpCall.java [hunk 3]
```diff
@@ -151,7 +151,7 @@ private static ResponseBody buffer(MediaType mediaType, byte[] bodyBytes) {
 
     @Override
     public void enqueue(Callback callback) {
-        DeferredTracer tracer = new DeferredTracer();
+        DeferredTracer tracer = new DeferredTracer("acquire-limiter");
```
Comment (carterkozak) [inline okhttp-clients/src/main/java/com/palantir/conjure/java/okhttp/RemotingOkHttpCall.java]: Would be nice if we prefixed this as well

File: okhttp-clients/src/test/java/com/palantir/conjure/java/okhttp/IoExceptionResponseHandlerTest.java [hunk 4]
```diff
@@ -51,7 +51,7 @@ public void extractsIoExceptionForAllErrorCodes() {
         for (int code : ImmutableList.of(300, 400, 404, 500)) {
             IOException exception = decode(MediaType.APPLICATION_JSON, code, "body").get();
             assertThat(exception.getMessage()).isEqualTo(
```
Comment (carterkozak) [inline okhttp-clients/src/test/java/com/palantir/conjure/java/okhttp/IoExceptionResponseHandlerTest.java:53]: nit: could use `assertThatLoggableExceptionThrownBy` here, even though safe-logging wasn't introduced by your change.

File: okhttp-clients/src/main/java/com/palantir/conjure/java/okhttp/IoExceptionResponseHandler.java (hunk 1 above)
Comment (carterkozak) [reply okhttp-clients/src/main/java/com/palantir/conjure/java/okhttp/IoExceptionResponseHandler.java:47]: 👍

File: okhttp-clients/src/main/java/com/palantir/conjure/java/okhttp/RemotingOkHttpCall.java (hunk 3 above)
Comment (pkoenig10) [reply okhttp-clients/src/main/java/com/palantir/conjure/java/okhttp/RemotingOkHttpCall.java]: What do you want to prefix this with? `OkHttp: `?

Comment (carterkozak) [reply okhttp-clients/src/main/java/com/palantir/conjure/java/okhttp/RemotingOkHttpCall.java]: Yep, we should group by component, not necessarily class

File: okhttp-clients/src/test/java/com/palantir/conjure/java/okhttp/IoExceptionResponseHandlerTest.java (hunk 4 above)
Comment (pkoenig10) [reply okhttp-clients/src/test/java/com/palantir/conjure/java/okhttp/IoExceptionResponseHandlerTest.java:53]: We're not throwing an exception here and the `decode` method returns a regular `IOException`. So we can either create a callable that throws this exception or cast it to a `SafeIoException` but both of these options seems a bit gross. Happy to do this if you would prefer though.

**Comment by carterkozak** [comment]
lgtm 👍

File: okhttp-clients/src/main/java/com/palantir/conjure/java/okhttp/RemotingOkHttpCall.java [hunk 5]
```diff
@@ -151,7 +151,7 @@ private static ResponseBody buffer(MediaType mediaType, byte[] bodyBytes) {
 
     @Override
     public void enqueue(Callback callback) {
-        DeferredTracer tracer = new DeferredTracer();
+        DeferredTracer tracer = new DeferredTracer("OkHttp: acquire-limiter");
```
Comment (robert3005) [inline okhttp-clients/src/main/java/com/palantir/conjure/java/okhttp/RemotingOkHttpCall.java]: I don't think this name is accurate as you need to open span now and close it in the success while this only make sure that there's same trace around retries

Comment (pkoenig10) [reply okhttp-clients/src/main/java/com/palantir/conjure/java/okhttp/RemotingOkHttpCall.java]: What alternative name would you suggest?

File: okhttp-clients/src/main/java/com/palantir/conjure/java/okhttp/RemotingOkHttpCall.java [hunk 6]
```diff
@@ -151,7 +151,7 @@ private static ResponseBody buffer(MediaType mediaType, byte[] bodyBytes) {
 
     @Override
     public void enqueue(Callback callback) {
-        DeferredTracer tracer = new DeferredTracer("OkHttp: acquire-limiter");
+        DeferredTracer tracer = new DeferredTracer("OkHttp: enqueue");
```
Comment (carterkozak) [inline okhttp-clients/src/main/java/com/palantir/conjure/java/okhttp/RemotingOkHttpCall.java:154]: @robert3005 Are you happy with this?

Comment (robert3005) [reply okhttp-clients/src/main/java/com/palantir/conjure/java/okhttp/RemotingOkHttpCall.java:154]: LGTM

**Comment by robert3005** [comment]
👍

**Comment by abless** [comment]
Note that this broke the Salt client: https://github.palantir.build/foundry/salt/blob/develop/salt-client/src/main/java/com/palantir/salt/client/Salt.java#L1368

File: okhttp-clients/src/main/java/com/palantir/conjure/java/okhttp/IoExceptionResponseHandler.java (hunk 1 above)
Comment (abless) [reply okhttp-clients/src/main/java/com/palantir/conjure/java/okhttp/IoExceptionResponseHandler.java:47]: This broke clients that previously tried parsing the error code.

**Pull request merged**